### PR TITLE
Add typing to ``Name._infer`` and ``AssignName.infer_lhs``

### PR DESCRIPTION
**Pull request by DanielNoord**
## Steps

- [x] Write a good description on what the PR does.

## Description

Doing one at a time...

## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :hammer: Refactoring   |

## Related Issue

### CODE REVIEW

File: astroid/bases.py [hunk 1]
```diff
@@ -128,7 +128,7 @@ def infer(
 def _infer_stmts(
     stmts: Sequence[nodes.NodeNG | type[Uninferable] | Instance],
     context: InferenceContext | None,
-    frame: nodes.NodeNG | Instance | None = None,
+    frame: nodes.NodeNG | Instance | str | None = None,
```
Comment (cdce8p) [inline astroid/bases.py:131]:
```suggestion
    frame: nodes.NodeNG | Instance | None = None,
```
See comment regarding `self.lookup` below.

File: astroid/inference.py [hunk 2]
```diff
@@ -197,15 +201,19 @@ def _higher_function_scope(node):
...
 
 
-def infer_name(self, context=None):
+def infer_name(
+    self: nodes.Name | nodes.AssignName,
+    context: InferenceContext | None = None,
+    **kwargs: Any,
+) -> Generator[InferenceResult, None, None]:
     """infer a Name: use name lookup rules"""
     frame, stmts = self.lookup(self.name)
```
Comment (cdce8p) [inline astroid/inference.py:218]: I believe the return type for `self.lookup` might be wrong. It returns a generator from `scope_lookup`. However, this does return a `NodeNG` as first tuple argument (usually `self`) and not a string.

File: astroid/inference.py [hunk 3]
```diff
@@ -185,7 +189,7 @@ def _infer_map(
 nodes.Dict._infer = infer_map  # type: ignore[assignment]
 
 
-def _higher_function_scope(node):
+def _higher_function_scope(node: nodes.LocalsDictNodeNG) -> nodes.FunctionDef | None:
```
Comment (cdce8p) [inline astroid/inference.py:192]: Why not annotate `node` itself with `NodeNG`? No attribute from `LocalsDictNodeNG` is used here.

**Review by cdce8p — CHANGES_REQUESTED** [review]

File: astroid/nodes/node_classes.py [hunk 4]
```diff
@@ -47,6 +47,13 @@
     from astroid.decorators import cachedproperty as cached_property
 
 
+class _InferLHS(Protocol):
+    def __call__(  # pylint: disable=no-self-argument
+        _self, self: AssignName, context: InferenceContext | None, **kwargs: Any
+    ) -> Generator[InferenceResult, None, None]:
+        ...
```
[the remark this answers is not in the document]
Comment (DanielNoord) [reply astroid/nodes/node_classes.py:54]: `pyright` kept complaining about those. I'll have another look and try to get it to work.

[the remark this answers is not in the document]
Comment (DanielNoord) [reply astroid/nodes/node_classes.py:54]: But how do you type the `**kwargs: any` then? Isn't that required as well?

[the remark this answers is not in the document]
Comment (cdce8p) [reply astroid/nodes/node_classes.py:54]: Luckily, it's only called once with `context`. So we don't actually need `**kwargs` 😅
https://github.com/PyCQA/astroid/blob/171b222b9ed1477b50f3e94c4266f97a67ed1c99/astroid/inference.py#L945

[the remark this answers is not in the document]
Comment (DanielNoord) [reply astroid/nodes/node_classes.py:54]: But `Name._infer` can be, right? And they share the same definition/signature?

File: astroid/inference.py (hunk 2 above)
Comment (DanielNoord) [reply astroid/inference.py:218]: Introduced in https://github.com/PyCQA/astroid/pull/1503/.

Never trust typing in `astroid` docstrings 😄

File: astroid/nodes/node_classes.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (cdce8p) [reply astroid/nodes/node_classes.py:54]: > But `Name._infer` can be, right? And they share the same definition/signature?

True, but the type annotation for `infer_lhs` doesn't need `**kwargs`. The assignment still works without them and as long as the single `infer_lhs` call doesn't pass any additional keyword arguments, everything is fine.

[the remark this answers is not in the document]
Comment (DanielNoord) [reply astroid/nodes/node_classes.py:54]: Ah I didn't know that. I'll use your suggestion then.

**Review by cdce8p — APPROVED** [review]

**Pull request merged**